### PR TITLE
feat: Switch to zod for the product API and implement missing primitives

### PR DESCRIPTION
**Pull request by sradevski**

### CODE REVIEW

File: integration-tests/api/__tests__/admin/product.js [hunk 1]
```diff
@@ -2764,10 +2769,10 @@ medusaIntegrationTestRunner({
 
           const payload = {
             title: "Second variant",
-            sku: variant.sku,
```
Comment (sradevski) [inline integration-tests/api/__tests__/admin/product.js:2767]: These were all null before which is not a valid value

File: packages/medusa/src/api-v2/admin/products/helpers.ts [hunk 2]
```diff
@@ -39,14 +41,27 @@ export const remapKeysForVariant = (selectFields: string[]) => {
...
+  TReq extends AdminCreateProductType | AdminUpdateProductType,
+  TResp extends CreateProductDTO | UpdateProductDTO
+>(
+  p: TReq
+): TResp => {
   return {
     ...p,
-    variants: p.variants?.map(remapVariant),
+    // TODO: Map to enum
+    status: p.status as ProductStatus,
```
Comment (sradevski) [inline packages/medusa/src/api-v2/admin/products/helpers.ts]: I will tackle this now

File: packages/medusa/src/api-v2/utils/validate-query.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Omit the non filterable config from the validated object
+ * @param obj
+ */
+const getFilterableFields = <T extends RequestQueryFields>(obj: T): T => {
+  const result = omit(obj, ["limit", "offset", "fields", "order"]) as T
+  return removeUndefinedProperties(result)
+}
+
+export function validateAndTransformQuery<TEntity extends BaseEntity>(
```
Comment (adrien2p) [inline packages/medusa/src/api-v2/utils/validate-query.ts:37]: q: shouldn't this be in the. middlewares?

File: packages/medusa/src/api-v2/admin/products/query-config.ts [hunk 4]
```diff
@@ -28,8 +28,6 @@ export const defaultAdminProductsVariantFields = [
 
 export const retrieveVariantConfig = {
   defaultFields: defaultAdminProductsVariantFields,
```
Comment (adrien2p) [inline packages/medusa/src/api-v2/admin/products/query-config.ts]: todo: defaults

File: packages/medusa/src/api-v2/utils/validators.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
+import { z } from "zod"
+
+export const createSelectParams = () => {
```
Comment (adrien2p) [inline packages/medusa/src/api-v2/utils/validators.ts:3]: waiting for those for my pr :D

**Review by adrien2p — APPROVED** [review]

File: packages/medusa/src/api-v2/admin/products/helpers.ts (hunk 2 above)
Comment (adrien2p) [reply packages/medusa/src/api-v2/admin/products/helpers.ts]: we can use nativeEnum for that

File: packages/medusa/src/api-v2/admin/products/helpers.ts [hunk 6]
```diff
@@ -39,14 +41,27 @@ export const remapKeysForVariant = (selectFields: string[]) => {
...
   }
 }
 
-export const remapVariant = (v: ProductVariantDTO) => {
+export const remapProductResponse = (p: ProductDTO) => {
+  return {
+    ...p,
```
Comment (adrien2p) [inline packages/medusa/src/api-v2/admin/products/helpers.ts]: todo: if we could remove all one letter please

**Pull request merged**